src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -8,16 +8,140 @@
+            => _channel.Writer.TryComplete();
+
+        /// <summary>
+        /// Signals that no further elements will be enqueued.  All outstanding and future
+        /// Dequeue Task will be cancelled.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is already completed.</exception>
+        public void Complete()
+        {
+            if (!TryComplete())
+                throw new InvalidOperationException($"Cannot call {nameof(Complete)} when the queue is already completed.");
+        }
+
+        public void PromiseNotToEnqueue()
+            => Volatile.Write(ref _promisedNotToEnqueue, true);
+
+        /// <summary>
+        /// Gets a task whose result is the element at the head of the queue. If the queue
+        /// is empty, the returned task waits for an element to be enqueued. If <see cref="Complete"/> 
+        /// is called before an element becomes available, the returned task is completed and
+        /// <see cref="Optional{T}.HasValue"/> will be <see langword="false"/>.
+        /// </summary>
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
+        public async ValueTask<Optional<TElement>> TryDequeueAsync(CancellationToken cancellationToken)
+        {
+            while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (_channel.Reader.TryRead(out var element))
+                    return new Optional<TElement>(element);
+
+                // someone else may have beat us to reading the element.  try again.
+            }
+
+            // Reader is entirely done.  Definitely can't dequeue anything at this point.
+            return default;
+        }
+
+        /// <summary>
+        /// Gets a task whose result is the element at the head of the queue. If the queue
+        /// is empty, the returned task waits for an element to be enqueued. If <see cref="Complete"/> 
+        /// is called before an element becomes available, the returned task is cancelled.
+        /// </summary>
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
+        public async Task<TElement> DequeueAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var result = await TryDequeueAsync(cancellationToken).ConfigureAwait(false);
+            if (result.HasValue)
+                return result.Value;
+
+            new CancellationToken(canceled: true).ThrowIfCancellationRequested();
+            throw ExceptionUtilities.Unreachable();
+        }
+    }
+
     /// <summary>
     /// A queue whose enqueue and dequeue operations can be performed in parallel.
     /// </summary>
     /// <typeparam name="TElement">The type of values kept by the queue.</typeparam>
-    internal sealed class AsyncQueue<TElement>
+    internal sealed class AsyncQueue1<TElement>
```

Comment:
keeping htis guy around just so i can have the source-code close by to compare.  Would remove if we went forward with this.